CE-919 - Update "Enable Auth0 single sign-on (SSO)" (#1421)
hashicorp/web-unified-docs: content/consul/v1.21.x/content/docs/secure/sso/auth0.mdx

Reviewer: ```suggestion
This functions associates your user and app metadata in Auth0 with Consul's data.
```

1. [Login with OIDC](#login-with-oidc)

## Configure Auth0 application

In order to configure an Auth0 application for Consul, you will first need to take note of your Consul agent address. This is the address that you use to access Consul, either via the UI or the CLI. To fetch it, run the following command:

```shell-session
$ echo $CONSUL_HTTP_ADDR
http://localhost:8500
```

In this case, the Consul agent address is `localhost`, running on the default port `8500`. Next, open the **Settings** tab in the [Auth0 App dashboard](https://manage.auth0.com/#/).

![Auth0 UI configure application to set callback URLs](/img/sso/auth0-configure-callback-urls.png)

In the **Allowed Callback URLs** field, enter the following:

```uri
http://localhost:8550/oidc/callback,
http://localhost:8500/ui/oidc/callback
```

The `http://localhost:8550/oidc/callback` address is for logins with the CLI `consul login -type=oidc -method=oidc` command.
The `http://localhost:8500/ui/oidc/callback` is for logins with the Consul UI.

<Note>

 The callback URLs must be comma-separated.

</Note>

## Create Auth0 users and metadata

> In the Auth0 dashboard for your app, on the sidebar, select **Users & Roles > Users** and click on _Create User_.

Reviewer: ```suggestion
In the Auth0 dashboard for your app, on the sidebar, select **Users & Roles > Users** and click on **Create User**.
```

# Enable Auth0 single sign-on (SSO)

This page describes the process to configure Consul to use Auth0 for single sign-on (SSO) in the Consul UI and CLI using [OIDC](https://en.wikipedia.org/wiki/OpenID_Connect).

OIDC authentication is useful when you want to deploy SSO widely in your organization and do not want to manage access with individual Consul ACL tokens. Once implemented, SSO will enable an interactive login procedure that can be initiated from either the Consul UI or the command line.

> In this example you will configure [Auth0](https://auth0.com/) as an identity provider, and use the user metadata in Auth0 to automatically grant permissions in Consul ACL. This approach unifies the user permissions grant without the need to create individual user tokens or to define permissions individually for the different users.

Reviewer: ```suggestion
These instructions configure Auth0 as an identity provider, so that Consul can use data in Auth0 to automatically grant permissions for Consul ACL. This approach unifies the user permissions grants without creating individual user tokens or manually defining individual permissions for different users.
```

<Note title="Enterprise Only">

 The single sign-on (SSO) functionality demonstrated here requires self-managed [Consul Enterprise](https://www.hashicorp.com/products/consul/pricing/). If you've purchased or wish to try out Consul Enterprise, refer to [how to access Consul Enterprise](/consul/docs/enterprise#access-consul-enterprise).

</Note>

## Workflow

1. [Configure an Auth0 application for Consul integration](#configure-auth0-application)

> 1. [Create Auth0 users and metadata](#create-auth0-users-and-metadata)

Reviewer: ```suggestion
1. [Create a rule to associate Auth0 users and Consul metadata](#create-auth0-users-and-metadata)
```
More descriptive for what the step entails, assuming a user already has Auth0 users to integrate.

http://<consul_agent_address>:8500/ui/oidc/callback
```

The `http://<consul_agent_address>:8550/oidc/callback` address will be used by the CLI when you login via `consul login -type=oidc -method=oidc` command.

> The `http://<consul_agent_address>:8500/ui/oidc/callback` address will be used by the Consul UI when you login with OIDC auth method.

Reviewer: ```suggestion
The `http://<consul_agent_address>:8500/ui/oidc/callback` is for logins with the Consul UI.
```

> ![Auth0 UI configure user metadata](/img/sso/auth0-configure-user-metadata.png)

Reviewer: ````suggestion
The data you associate with Auth0 may change depending on your existing configurations. This example uses `user_metadata` and `app_metadata` to configure Auth0.

```json
{
// user_metadata (end-user editable)
"first_name": "Kara",
"last_name": "Danvers"
}
```

```json
{
// app_metadata (end-user not editable)
"roles": ["engineering"]
}
```
````

> Then, in the Auth0 dashboard for your application, on the sidebar, select **Rules** and create a new rule with the following content:

Reviewer: ```suggestion
In Auth0, create a new rule with the following javascript function:
```

}
```

This functions associates your user and app metadata in Auth0 with Consul's data.

This example uses `http://consul.internal` to namespace the claims that are not part of the [JWT RFC](https://tools.ietf.org/html/rfc7519). The namespace is arbitrary, but it must be unique. Auth0 enforces the presence of a unique namespace by discarding claims that are unnamespaced and not in the RFC.

## Enable the OIDC auth method for Consul

> In order to enable the OIDC auth method for Consul, you need to create a new auth method configuration that points to your Auth0 application. The contents of this configuration file depend on which way you supply it to Consul. You can either use the Consul CLI or the HTTP API.

Reviewer: ```suggestion
In order to enable the OIDC auth method for Consul, you need to create a new auth method configuration that points to your Auth0 application. The contents of this configuration file depend on which way you supply it to Consul. 

You can either use the Consul CLI or the HTTP API. Note that the configuration file contents differ slightly depending on whether you use the CLI or the HTTP API.
```

</Tab>
</Tabs>

Once you have configured the auth method, you can automate permissions grants to users using the metadata you defined earlier. This means that once auth method trust is established, Consul can be configured to bind attested identities to roles or services with no additional work beyond what is required to link the identity and the auth method. This can be configured using Consul binding rules.

### Grant role permissions with app metadata

> Assign the Consul role `eng-ro` to users in the Auth0 `engineering` group.

Reviewer: ```suggestion
Next, create a Consul ACL role and token to associate with the Auth0 method. The following command creates an ACL role named `eng-ro` and assigns it to users in the Auth0 `engineering` group.
```

Selector:     engineering in list.groups
```

> This will automatically associate every user with `engineering` in their `app_metadata` in Auth0 to the Consul `eng-ro` role. Check the following documentation for further information on Consul [roles](consul/docs/secure/acl/role) and Consul [policies](/consul/docs/secure/acl/policy).

Reviewer: ```suggestion
Every user with `engineering` in their `app_metadata` in Auth0 will be assigned the Consul `eng-ro` role. Check the following documentation for further information on [Consul ACL roles](consul/docs/secure/acl/role) and [Consul ACL policies](/consul/docs/secure/acl/policy).
```

### Grant service permissions with user metadata

> If you want to grant users permissions to register services, you can use a service type binding. Here's how to grant users in the `engineering` group the ability to register a service in the service mesh with their own name.

Reviewer: ```suggestion
If you want to grant users permissions to register services, use a service type binding. The following example grants users in the `engineering` group the ability to register a service using a token that is dynamically generated from their own name.
```

> In this tutorial you learned how to enable SSO for Consul Enterprise, and how to use Auth0 as the OIDC provider.

Reviewer: ```suggestion
## Next steps
```
